File: source/daemon/crates/wardnet-types/src/tests/dhcp.rs
use std::net::Ipv4Addr;

use crate::dhcp::{
    DhcpConfig, DhcpLease, DhcpLeaseEventType, DhcpLeaseLog, DhcpLeaseStatus, DhcpReservation,
};
use uuid::Uuid;

#[test]
fn lease_status_round_trip() {
    for status in [
        DhcpLeaseStatus::Active,
        DhcpLeaseStatus::Expired,
        DhcpLeaseStatus::Released,
    ] {
        let json = serde_json::to_string(&status).unwrap();
        let back: DhcpLeaseStatus = serde_json::from_str(&json).unwrap();
        assert_eq!(status, back);
    }
}

#[test]
fn lease_status_snake_case_rename() {
    assert_eq!(
        serde_json::to_string(&DhcpLeaseStatus::Active).unwrap(),
        "\"active\""
    );
    assert_eq!(
        serde_json::to_string(&DhcpLeaseStatus::Expired).unwrap(),
        "\"expired\""
    );
    assert_eq!(
        serde_json::to_string(&DhcpLeaseStatus::Released).unwrap(),
        "\"released\""
    );
}

#[test]
fn lease_event_type_round_trip() {
    for event_type in [
        DhcpLeaseEventType::Assigned,
        DhcpLeaseEventType::Renewed,
        DhcpLeaseEventType::Released,
        DhcpLeaseEventType::Expired,
        DhcpLeaseEventType::Conflict,
    ] {
        let json = serde_json::to_string(&event_type).unwrap();
        let back: DhcpLeaseEventType = serde_json::from_str(&json).unwrap();
        assert_eq!(event_type, back);
    }
}

#[test]
fn lease_event_type_snake_case_rename() {
    assert_eq!(
        serde_json::to_string(&DhcpLeaseEventType::Assigned).unwrap(),
        "\"assigned\""
    );
    assert_eq!(
        serde_json::to_string(&DhcpLeaseEventType::Conflict).unwrap(),
        "\"conflict\""
    );
}

#[test]
fn dhcp_lease_round_trip() {
    let lease = DhcpLease {
        id: Uuid::parse_str("00000000-0000-0000-0000-000000000001").unwrap(),
        mac_address: "AA:BB:CC:DD:EE:01".to_owned(),
        ip_address: Ipv4Addr::new(192, 168, 1, 100),
        hostname: Some("myphone".to_owned()),
        lease_start: "2026-03-07T00:00:00Z".parse().unwrap(),
        lease_end: "2026-03-08T00:00:00Z".parse().unwrap(),
        status: DhcpLeaseStatus::Active,
        device_id: None,
        created_at: "2026-03-07T00:00:00Z".parse().unwrap(),
        updated_at: "2026-03-07T00:00:00Z".parse().unwrap(),
    };
    let json = serde_json::to_string(&lease).unwrap();
    let back: DhcpLease = serde_json::from_str(&json).unwrap();
    assert_eq!(lease.id, back.id);
    assert_eq!(lease.mac_address, back.mac_address);
    assert_eq!(lease.ip_address, back.ip_address);
    assert_eq!(lease.hostname, back.hostname);
    assert_eq!(lease.status, back.status);
    assert_eq!(lease.lease_start, back.lease_start);
    assert_eq!(lease.lease_end, back.lease_end);
}

#[test]
fn dhcp_reservation_round_trip() {
    let reservation = DhcpReservation {
        id: Uuid::parse_str("00000000-0000-0000-0000-000000000002").unwrap(),
        mac_address: "AA:BB:CC:DD:EE:02".to_owned(),
        ip_address: Ipv4Addr::new(192, 168, 1, 50),
        hostname: Some("printer".to_owned()),
        description: Some("Office printer".to_owned()),
        created_at: "2026-03-07T00:00:00Z".parse().unwrap(),
        updated_at: "2026-03-07T00:00:00Z".parse().unwrap(),
    };
    let json = serde_json::to_string(&reservation).unwrap();
    let back: DhcpReservation = serde_json::from_str(&json).unwrap();
    assert_eq!(reservation.id, back.id);
    assert_eq!(reservation.mac_address, back.mac_address);
    assert_eq!(reservation.ip_address, back.ip_address);
    assert_eq!(reservation.hostname, back.hostname);
    assert_eq!(reservation.description, back.description);
}

#[test]
fn dhcp_config_round_trip() {
    let config = DhcpConfig {
        enabled: true,
        pool_start: Ipv4Addr::new(192, 168, 1, 100),
        pool_end: Ipv4Addr::new(192, 168, 1, 200),
        subnet_mask: Ipv4Addr::new(255, 255, 255, 0),
        upstream_dns: vec![Ipv4Addr::new(1, 1, 1, 1), Ipv4Addr::new(8, 8, 8, 8)],
        lease_duration_secs: 86400,
        router_ip: Some(Ipv4Addr::new(192, 168, 1, 1)),
    };
    let json = serde_json::to_string(&config).unwrap();
    let back: DhcpConfig = serde_json::from_str(&json).unwrap();
    assert_eq!(config.enabled, back.enabled);
    assert_eq!(config.pool_start, back.pool_start);
    assert_eq!(config.pool_end, back.pool_end);
    assert_eq!(config.subnet_mask, back.subnet_mask);
    assert_eq!(config.upstream_dns, back.upstream_dns);
    assert_eq!(config.lease_duration_secs, back.lease_duration_secs);
    assert_eq!(config.router_ip, back.router_ip);
}

#[test]
fn dhcp_lease_log_round_trip() {
    let log = DhcpLeaseLog {
        id: 1,
        lease_id: Uuid::parse_str("00000000-0000-0000-0000-000000000001").unwrap(),
        event_type: DhcpLeaseEventType::Assigned,
        details: Some("Initial assignment".to_owned()),
        created_at: "2026-03-07T00:00:00Z".parse().unwrap(),
    };
    let json = serde_json::to_string(&log).unwrap();
    let back: DhcpLeaseLog = serde_json::from_str(&json).unwrap();
    assert_eq!(log.id, back.id);
    assert_eq!(log.lease_id, back.lease_id);
    assert_eq!(log.event_type, back.event_type);
}

File: source/sdk/wardnet-go/dhcp.go
package wardnet

import (
	"context"
	"time"

	"wardnet.network/go/internal/rest"
)

// DHCPService configures the DHCP server, its leases, and its reservations.
type DHCPService struct{ c *Client }

// DHCPConfig is the DHCP pool configuration. The gateway address is
// auto-detected by the daemon rather than configured.
type DHCPConfig struct {
	// Enabled reports whether the server should be running.
	Enabled bool `json:"enabled"`
	// GatewayIP is wardnet's own LAN address, advertised as gateway and DNS.
	GatewayIP string `json:"gateway_ip"`
	// RouterIP is the upstream router, offered as the secondary gateway.
	RouterIP string `json:"router_ip"`
	// PoolStart / PoolEnd bound the dynamic address range.
	PoolStart string `json:"pool_start"`
	PoolEnd   string `json:"pool_end"`
	// SubnetMask is the mask handed to clients.
	SubnetMask string `json:"subnet_mask"`
	// UpstreamDNS are the resolvers offered when wardnet's own is not used.
	UpstreamDNS []string `json:"upstream_dns"`
	// LeaseDurationSecs is how long an offered lease is valid.
	LeaseDurationSecs int32 `json:"lease_duration_secs"`
}

// DHCPServerStatus is the DHCP server's live state. It is named apart from
// [DHCPStatus], which describes how one device's address is managed.
type DHCPServerStatus struct {
	// Enabled is the configured intent; Running is the observed reality.
	Enabled bool `json:"enabled"`
	Running bool `json:"running"`
	// ActiveLeaseCount is how many leases are currently held.
	ActiveLeaseCount int64 `json:"active_lease_count"`
	// PoolUsed / PoolTotal describe address exhaustion.
	PoolUsed  int64 `json:"pool_used"`
	PoolTotal int64 `json:"pool_total"`
}

// Lease is an active or historical DHCP lease.
type Lease struct {
	// ID is the lease's UUID.
	ID string `json:"id"`
	// MAC is the client's hardware address.
	MAC string `json:"mac_address"`
	// IP is the leased address.
	IP string `json:"ip_address"`
	// Hostname is the name the client claimed, when it sent one.
	Hostname string `json:"hostname"`
	// DeviceID is the matching known device, when one is tracked.
	DeviceID string `json:"device_id"`
	// Status is "active", "expired", or "released".
	Status string `json:"status"`
	// LeaseStart / LeaseEnd bound the lease's validity.
	LeaseStart time.Time `json:"lease_start"`
	LeaseEnd   time.Time `json:"lease_end"`
}

// Reservation is a static MAC-to-IP binding.
type Reservation struct {
	// ID is the reservation's UUID.
	ID string `json:"id"`
	// MAC is the hardware address the address is pinned to.
	MAC string `json:"mac_address"`
	// IP is the reserved address.
	IP string `json:"ip_address"`
	// Hostname is an optional name to hand back to the client.
	Hostname string `json:"hostname"`
	// Description is an optional operator note.
	Description string `json:"description"`
	// CreatedAt is when the reservation was made.
	CreatedAt time.Time `json:"created_at"`
}

// CreateReservationParams describes a reservation to add.
type CreateReservationParams struct {
	// MAC is the hardware address to pin.
	MAC string
	// IP is the address to hand it.
	IP string
	// Hostname is an optional name for the client.
	Hostname string
	// Description is an optional operator note.
	Description string
}

// Config fetches the DHCP configuration.
func (s *DHCPService) Config(ctx context.Context) (*DHCPConfig, error) {
	resp, err := s.c.rest.DhcpGetConfigWithResponse(ctx)
	if err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		return nil, apiError(resp.HTTPResponse, resp.Body)
	}
	cfg := dhcpConfigFromREST(&resp.JSON200.Config)
	return &cfg, nil
}

// SetEnabled starts or stops the DHCP server and returns the resulting config.
func (s *DHCPService) SetEnabled(ctx context.Context, enabled bool) (*DHCPConfig, error) {
	resp, err := s.c.rest.DhcpToggleWithResponse(ctx, rest.DhcpToggleJSONRequestBody{Enabled: enabled})
	if err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		return nil, apiError(resp.HTTPResponse, resp.Body)
	}
	cfg := dhcpConfigFromREST(&resp.JSON200.Config)
	return &cfg, nil
}

// Status reports the DHCP server's live state.
func (s *DHCPService) Status(ctx context.Context) (*DHCPServerStatus, error) {
	resp, err := s.c.rest.DhcpStatusWithResponse(ctx)
	if err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		return nil, apiError(resp.HTTPResponse, resp.Body)
	}
	m := resp.JSON200
	return &DHCPServerStatus{
		Enabled:          m.Enabled,
		Running:          m.Running,
		ActiveLeaseCount: m.ActiveLeaseCount,
		PoolUsed:         m.PoolUsed,
		PoolTotal:        m.PoolTotal,
	}, nil
}

// Leases returns every known lease.
func (s *DHCPService) Leases(ctx context.Context) ([]Lease, error) {
	resp, err := s.c.rest.ListLeasesWithResponse(ctx)
	if err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		return nil, apiError(resp.HTTPResponse, resp.Body)
	}
	out := make([]Lease, 0, len(resp.JSON200.Leases))
	for i := range resp.JSON200.Leases {
		l := &resp.JSON200.Leases[i]
		lease := Lease{
			ID:         l.Id.String(),
			MAC:        l.MacAddress,
			IP:         l.IpAddress,
			Hostname:   deref(l.Hostname),
			Status:     string(l.Status),
			LeaseStart: l.LeaseStart,
			LeaseEnd:   l.LeaseEnd,
		}
		if l.DeviceId != nil {
			lease.DeviceID = l.DeviceId.String()
		}
		out = append(out, lease)
	}
	return out, nil
}

// RevokeLease releases a lease, freeing its address for reuse.
func (s *DHCPService) RevokeLease(ctx context.Context, id string) error {
	lid, err := parseUUID(id, "lease")
	if err != nil {
		return err
	}
	resp, err := s.c.rest.RevokeLeaseWithResponse(ctx, lid)
	if err != nil {
		return err
	}
	if resp.JSON200 == nil {
		return apiError(resp.HTTPResponse, resp.Body)
	}
	return nil
}

// Reservations returns every static MAC-to-IP binding.
func (s *DHCPService) Reservations(ctx context.Context) ([]Reservation, error) {
	resp, err := s.c.rest.ListReservationsWithResponse(ctx)
	if err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		return nil, apiError(resp.HTTPResponse, resp.Body)
	}
	out := make([]Reservation, 0, len(resp.JSON200.Reservations))
	for i := range resp.JSON200.Reservations {
		out = append(out, reservationFromREST(&resp.JSON200.Reservations[i]))
	}
	return out, nil
}

// AddReservation pins an address to a MAC.
func (s *DHCPService) AddReservation(ctx context.Context, params CreateReservationParams) (*Reservation, error) {
	body := rest.CreateReservationJSONRequestBody{
		MacAddress: params.MAC,
		IpAddress:  params.IP,
	}
	if params.Hostname != "" {
		body.Hostname = ptr(params.Hostname)
	}
	if params.Description != "" {
		body.Description = ptr(params.Description)
	}
	resp, err := s.c.rest.CreateReservationWithResponse(ctx, body)
	if err != nil {
		return nil, err
	}
	if resp.JSON201 == nil {
		return nil, apiError(resp.HTTPResponse, resp.Body)
	}
	r := reservationFromREST(&resp.JSON201.Reservation)
	return &r, nil
}

// RemoveReservation deletes a static MAC-to-IP binding.
func (s *DHCPService) RemoveReservation(ctx context.Context, id string) error {
	rid, err := parseUUID(id, "reservation")
	if err != nil {
		return err
	}
	resp, err := s.c.rest.DeleteReservationWithResponse(ctx, rid)
	if err != nil {
		return err
	}
	if resp.JSON200 == nil {
		return apiError(resp.HTTPResponse, resp.Body)
	}
	return nil
}

func dhcpConfigFromREST(c *rest.DhcpConfig) DHCPConfig {
	return DHCPConfig{
		Enabled:           c.Enabled,
		GatewayIP:         c.GatewayIp,
		RouterIP:          deref(c.RouterIp),
		PoolStart:         c.PoolStart,
		PoolEnd:           c.PoolEnd,
		SubnetMask:        c.SubnetMask,
		UpstreamDNS:       c.UpstreamDns,
		LeaseDurationSecs: c.LeaseDurationSecs,
	}
}

func reservationFromREST(r *rest.DhcpReservation) Reservation {
	return Reservation{
		ID:          r.Id.String(),
		MAC:         r.MacAddress,
		IP:          r.IpAddress,
		Hostname:    deref(r.Hostname),
		Description: deref(r.Description),
		CreatedAt:   r.CreatedAt,
	}
}
